### ISSUE TITLE
How to Convert a Merged Seurat Object with Multiple Samples to a Usable Semla Object?

### ISSUE THREAD

**Issue by yangxinzhi** [ludvigla/semla#40]
![image](https://github.com/user-attachments/assets/e53ebb29-5177-4d24-bb2b-ccc1b727197a)
![image](https://github.com/user-attachments/assets/ac0a4ea0-ad4f-4be2-9e13-1c2b4c9f1156)

Hello, I am encountering an issue when using the Semla package. Specifically, I am trying to convert my merged Seurat object (which contains multiple samples) into a format that is compatible with Semla, but I keep getting errors. Is there a way to make VisiumV2 data compatible with Semla?

**Comment by yangxinzhi** [comment]
![image](https://github.com/user-attachments/assets/4013c29f-c7f4-4223-b1aa-96997fc0a50a)
![image](https://github.com/user-attachments/assets/c78f538c-cf5e-419e-bd88-b1fb832994a5)
I even attempted to convert the V5 format to V4 for compatibility, but the same error persists.

**Comment by yangxinzhi** [comment]
![image](https://github.com/user-attachments/assets/2b31830b-1d3c-4bde-8dda-4c7eb001913a)

**Comment by jemorlanes** [comment]
Hello!

Are you working with VisiumHD data?

**Comment by yangxinzhi** [comment]
Hello!
I wanted to let you know that when I use Seurat version 5.1.0, I encountered an issue with image formatting. Specifically, when I used the following commands:

```R
Early2 <- CreateSeuratObject(counts = Read10X("Early_R2/"), assay = "Spatial") # Load expression matrix and create Seurat object
Early2_img <- Read10X_Image(image.dir = file.path("F:/博士单细胞分析/空间转录组/GSE212903_spatial/Young_R02_S1_spatial/Young_R02_S1_spatial/"), filter.matrix = TRUE) # Load a 10X Genomics Visium Image
Early2_img <- Early2_img[Cells(x = Early2)]
DefaultAssay(Early2_img) <- "Spatial"
```

This command creates a v2-format image object. However, when I use a modified function from GitHub, as shown below:

```R
Read10X_Image <- function(image.dir, image.name = "tissue_hires_image.png", filter.matrix = TRUE, ...) {
  
  library(png)
  library(jsonlite)
  
  image <- readPNG(source = file.path(image.dir, image.name))
  scale.factors <- fromJSON(txt = file.path(image.dir, 'scalefactors_json.json'))
  tissue.positions.path <- Sys.glob(paths = file.path(image.dir, 'tissue_positions*'))
  tissue.positions <- read.csv(
    file = tissue.positions.path[1],
    col.names = c('barcodes', 'tissue', 'row', 'col', 'imagerow', 'imagecol'),
    header = ifelse(
      test = basename(tissue.positions.path[1]) == "tissue_positions.csv",
      yes = TRUE,
      no = FALSE
    ),
    as.is = TRUE,
    row.names = 1
  )
  if (filter.matrix) {
    tissue.positions <- tissue.positions[which(x = tissue.positions$tissue == 1), , drop = FALSE]
  }
  unnormalized.radius <- scale.factors$fiducial_diameter_fullres * scale.factors$tissue_hires_scalef
  spot.radius <- unnormalized.radius / max(dim(x = image))
  return(new(
    Class = 'VisiumV1',
    image = image,
    scale.factors = scalefactors(
      spot = scale.factors$tissue_hires_scalef,
      fiducial = scale.factors$fiducial_diameter_fullres,
      hires = scale.factors$tissue_hires_scalef,
      scale.factors$tissue_hires_scalef
    ),
    coordinates = tissue.positions,
    spot.radius = spot.radius
  ))
}
```

This indeed outputs v1-format data. However, when I use this v1-format data with Seurat’s `SpatialDimPlot` function, as in:

```R
SpatialDimPlot(Spatial_integrated, images = 'Early2', image.alpha = 0, pt.size.factor = 3)
```

I find that parameters like `pt.size.factor = 3` don’t work. So I am currently only able to read the data using Seurat’s default `Read10X_Image` function to maintain functionality.

Hope this clarifies my issue!

**Comment by yangxinzhi** [comment]
Additionally, I found that when I attempted to convert the v2 format to v1 format using the following code:

```R
scale_factors <- scalefactors(
    spot = Early1_img@scale.factors$spot,
    fiducial = Early1_img@scale.factors$fiducial,
    hires = Early1_img@scale.factors$hires,
    lowres = Early1_img@scale.factors$lowres
)
coordinates_df <- as.data.frame(Early1_img@centroids@coords)

img_v1 <- new(
    Class = "VisiumV1",
    image = Early1_img@image,
    coordinates = coordinates_df,
    spot.radius = Early1_img@boundaries$centroids@radius,
    scale.factors = scale_factors
)
```

I was able to successfully convert it. However, when using `semla.data <- UpdateSeuratForSemla(Spatial_integrated)`, I encountered the following output:

```R
ℹ Found VisiumV1 object(s).
ℹ Expecting 'tissue_lowres' format.

── Collecting data from @images slot 
ℹ Collecting data for sample 1:
→   Exporting H&E image width dimensions 600x578 to C:\Users\Public\Documents\Wondershare\CreatorTemp\Rtmp8GmGRK/1.png
→   Collected scale factors.
→   Collected image information.
→   Collected coordinates.
ℹ Collecting data for sample 2:
→   Exporting H&E image width dimensions 600x577 to C:\Users\Public\Documents\Wondershare\CreatorTemp\Rtmp8GmGRK/2.png
→   Collected scale factors.
→   Collected image information.
→   Collected coordinates.
...
! Paths to raw images are unavailable. See '?ReplaceImagePaths()' for more information on how to update the paths.

✔ Returning updated Seurat object.
```

Then, when I ran the following commands to check if the paths were correctly set:

```R
lapply(semla.data@images, function(img) img@image.path)
```

I got the error:

```
Error in FUN(X[[i]], ...) : 
  no slot of name "image.path" for this object of class "VisiumV1"
```

Finally, when trying to analyze using Semla’s function with the following code:

```R
MapFeaturesSummary(semla.data, features = "nFeature_Spatial", subplot_type = "box")
```

I received the error:

```
Error in vapply(X = .FilterObjects(object = object, classes.keep = c("SpatialImage",  : 
  values must be length 1, but FUN(X[[6]]) result is length 0
```
I would like to know if there is currently a simple way to convert a Seurat object with multiple merged spatial transcriptomics samples into an object that can be used with Semla for improved visualization.

**Comment by yangxinzhi** [comment]
![image](https://github.com/user-attachments/assets/4fb24a92-d903-46cf-8b29-abb9b0802c17)
Is it because this can only analyze a Seurat object that contains only one image at a time, and then use MergeSTData for merging afterward?

**Comment by jemorlanes** [comment]
Hi! 

We have not fully migrated all of semla's methods to Seuratv5, which is from where this issue is coming from. Most of semla's methods work regardless, but some features might indeed not work. In previous versions from Seurat, there was no "VisiumV2". We will take a look at this specific case and see if we can make a quick fix for it!

Will reply to this thread once we figure it out, thank you for the comment! :))

**Comment by jemorlanes** [comment]
Hello! We have not merged into the main branch the updated version of `UpdateSeuratForSemla`, but we have a branch that has solved that issue.

Reinstall semla using `remotes::install_github("ludvigla/semla", ref = "upd_seuTosem")` and let us know if you it runs successfully! In the coming days we will merge this branch into `main`, but for now this should work. We will comment again once it is merged.
